From akpm@osdl.org Fri Feb 24 16:53:44 2006
Message-Id: <200602250053.k1P0rdJH016580@shell0.pdx.osdl.net>
Subject: USB: fix EHCI BIOS handshake
To: david-b@pacbell.net, dbrownell@users.sourceforge.net, greg@kroah.com,
   yazar256@gmail.com, mm-commits@vger.kernel.org
From: David Brownell <david-b@pacbell.net>
Date: Fri, 24 Feb 2006 16:55:52 -0800

From: David Brownell <david-b@pacbell.net>

Fix http://bugzilla.kernel.org/show_bug.cgi?id=6128

Finish morphing the "early handoff" version of the EHCI BIOS handshake over
to match the previous implementation inside the EHCI driver (except that
now we forcibly disable the SMI).  The version that had been with the PCI
code was surprisingly full of bugs.

Signed-off-by: David Brownell <dbrownell@users.sourceforge.net>
Cc: <yazar256@gmail.com>
Signed-off-by: Andrew Morton <akpm@osdl.org>
Signed-off-by: Greg Kroah-Hartman <gregkh@suse.de>
---
 drivers/usb/host/pci-quirks.c |   11 ++++++-----
 1 file changed, 6 insertions(+), 5 deletions(-)

--- gregkh-2.6.orig/drivers/usb/host/pci-quirks.c
+++ gregkh-2.6/drivers/usb/host/pci-quirks.c
@@ -260,12 +260,13 @@ static void __devinit quirk_usb_disable_
 						offset + EHCI_USBLEGCTLSTS,
 						val | EHCI_USBLEGCTLSTS_SOOE);
 #endif
-			}
 
-			/* always say Linux will own the hardware
-			 * by setting EHCI_USBLEGSUP_OS.
-			 */
-			pci_write_config_byte(pdev, offset + 3, 1);
+				/* some systems get upset if this semaphore is
+				 * set for any other reason than forcing a BIOS
+				 * handoff..
+				 */
+				pci_write_config_byte(pdev, offset + 3, 1);
+			}
 
 			/* if boot firmware now owns EHCI, spin till
 			 * it hands it over.
